src/coreclr/jit/llvm.cpp

```diff
@@ -1301,6 +1247,15 @@ void Llvm::fillPhis()
             auto iter = _localsMap->find({ lclNum, ssaNum });
             if (iter == _localsMap->end())
             {
+                //TODO-LLVM: if a phi arg is an arg that is not in the LLVM args, then we may not have a local for it
+                // For now as this PR is already big, just fail these functions.
+                // For a fix, we could do another lowering pass after SSA, copying
+                // those args into locals.  We may have to do that anyway to get the address of locals, although that might
+                // be ok to do before SSA
+                if (!(_compiler->lvaIsParameter(lclNum) && ssaNum == SsaConfig::FIRST_SSA_NUM))
+                {
+                    failFunctionCompilation();
+                }
```

Comment:
What are the scenarios that fail here?

One I can think of is uninitialized locals ("use before def locals"), it seems overkill to replace them "whole", inserting a "fake" def (say to zero for a start, though I believe LLVM provides more sophisticated mechanisms to indicate a local is uninitialized) in the first block seems enough.

Reply:
`IntPtr__CompareTo` is the method.

`t194` is the problem here.  There is no `d` for 2:1 and all the predecessor are above it.  When I wrote the comment I thought it was a function argument, but its not, so I'm not sure what's gone wrong.
```
------------ BB01 [???..???), preds={} succs={BB02}

------------ BB02 [000..009) -> BB04 (cond), preds={BB01} succs={BB03,BB04}
               [000100] ------------                 IL_OFFSET void   IL offset: 0x0
               [000101] ------------                 IL_OFFSET void   IL offset: 0x1
               [000126] ------------       t126 =    CNS_INT   int    4
               [000127] ------------       t127 =    LCL_VAR   int    V12 tmp4         u:1
                                                  /--*  t127   int
                                                  +--*  t126   int
               [000128] ------------       t128 = *  ADD       int
                                                  /--*  t128   int
N001 (  3,  2) [000002] ------------         t2 = *  IND       ref
N002 (  1,  1) [000003] ------------         t3 =    CNS_INT   ref    null
                                                  /--*  t2     ref
                                                  +--*  t3     ref
N003 (  8,  4) [000004] ------------         t4 = *  EQ        int
                                                  /--*  t4     int
N005 ( 12,  7) [000006] DA----------              *  STORE_LCL_VAR int    V03 loc1         d:1
               [000102] ------------                 IL_OFFSET void   IL offset: 0x6
N001 (  3,  2) [000007] ------------         t7 =    LCL_VAR   int    V03 loc1         u:1 (last use)
N002 (  1,  1) [000008] ------------         t8 =    CNS_INT   int    0
                                                  /--*  t7     int
                                                  +--*  t8     int
N003 (  5,  4) [000009] J------N----         t9 = *  EQ        int
                                                  /--*  t9     int
N004 (  7,  6) [000010] ------------              *  JTRUE     void

------------ BB03 [009..00E) -> BB14 (always), preds={BB02} succs={BB14}
               [000103] ------------                 IL_OFFSET void   IL offset: 0x9
               [000104] ------------                 IL_OFFSET void   IL offset: 0xa
N001 (  1,  1) [000087] ------------        t87 =    CNS_INT   int    1
                                                  /--*  t87    int
N003 (  5,  4) [000089] DA----------              *  STORE_LCL_VAR int    V04 loc2         d:5
               [000105] ------------                 IL_OFFSET void   IL offset: 0xc

------------ BB04 [00E..016) -> BB06 (cond), preds={BB02} succs={BB05,BB06}
               [000106] ------------                 IL_OFFSET void   IL offset: 0xe
               [000129] ------------       t129 =    CNS_INT   int    4
               [000130] ------------       t130 =    LCL_VAR   int    V12 tmp4         u:1
                                                  /--*  t130   int
                                                  +--*  t129   int
               [000131] ------------       t131 = *  ADD       int
                                                  /--*  t131   int
N003 (  3,  2) [000011] ------------        t11 = *  IND       ref
N004 (  1,  4) [000012] H------N----        t12 =    CNS_INT(h) int    0x420050 class
                                                  /--*  t11    ref    arg1 in rdx
                                                  +--*  t12    int    arg0 in rcx
N005 ( 18, 10) [000013] --C-G-------        t13 = *  CALL help ref    HELPER.CORINFO_HELP_ISINSTANCEOFCLASS
N006 (  1,  1) [000014] ------------        t14 =    CNS_INT   ref    null
                                                  /--*  t13    ref
                                                  +--*  t14    ref
N007 ( 20, 12) [000015] J---G--N----        t15 = *  EQ        int
                                                  /--*  t15    int
N008 ( 22, 14) [000016] ----G-------              *  JTRUE     void

------------ BB05 [016..020) -> BB07 (always), preds={BB04} succs={BB07}
               [000107] ------------                 IL_OFFSET void   IL offset: 0x16
               [000132] ------------       t132 =    CNS_INT   int    4
               [000133] ------------       t133 =    LCL_VAR   int    V12 tmp4         u:1
                                                  /--*  t133   int
                                                  +--*  t132   int
               [000134] ------------       t134 = *  ADD       int
                                                  /--*  t134   int
N003 (  3,  2) [000075] ------------        t75 = *  IND       ref
N004 (  1,  4) [000076] H-----------        t76 =    CNS_INT(h) int    0x420050 class
                                                  /--*  t75    ref    arg1 in rdx
                                                  +--*  t76    int    arg0 in rcx
N005 ( 18, 10) [000077] --CXG-------        t77 = *  CALL help byref  HELPER.CORINFO_HELP_UNBOX
                                                  /--*  t77    byref
N006 ( 21, 12) [000078] *--XG-------        t78 = *  IND       int
                                                  /--*  t78    int
N008 ( 25, 15) [000080] DA-XG-------              *  STORE_LCL_VAR int    V02 loc0         d:3
               [000108] ------------                 IL_OFFSET void   IL offset: 0x1d
N001 (  1,  1) [000081] ------------        t81 =    CNS_INT   int    1
                                                  /--*  t81    int
N003 (  5,  4) [000084] DA----------              *  STORE_LCL_VAR int    V09 tmp1         d:3

------------ BB06 [020..021), preds={BB04} succs={BB07}
               [000109] ------------                 IL_OFFSET void   IL offset: 0x20
N001 (  1,  1) [000017] ------------        t17 =    CNS_INT   int    0
                                                  /--*  t17    int
N003 (  5,  4) [000019] DA----------              *  STORE_LCL_VAR int    V09 tmp1         d:2

------------ BB07 [021..025) -> BB13 (cond), preds={BB05,BB06} succs={BB08,BB13}
     (  0,  0) [000196] ------------       t196 =    PHI_ARG   int    V02 loc0         u:3
     (  0,  0) [000195] ------------       t195 =    PHI_ARG   int    V09 tmp1         u:3
     (  0,  0) [000194] ------------       t194 =    PHI_ARG   int    V02 loc0         u:1
     (  0,  0) [000193] ------------       t193 =    PHI_ARG   int    V09 tmp1         u:2
```

Reply:
Could you share the full dump? FWIW, it looks like an "uninitialized local" case to me looking at this IR (and in this case I think the PHI is also "dead", though that will not always be the case and it is something we should tolerate).

So, there are two cases: actual uninitialized locals and those that are must-init (with zero). The logic to decide which is which [is not straightforward](https://github.com/dotnet/runtime/blob/03e0187a1a912f38e718c00614201dd20fa1bff1/src/coreclr/jit/codegencommon.cpp#L4616)...

So I agree we should leave this as-is for this PR (maybe rephrase the comment though), and look at it separately. Likely we'll need to insert explicit zero-initalization (and copy (an approximation of) the logic above) after SSA (because SSA already accounts for these rules and does some optimizations based on them). Then for true uninitialized locals we can use something like LLVM's `freeze undef`.

Reply:
Thanks, have updated the comment with the full dump.  Call args appear as `wasm arg` as there's a minor problem with dumping the tree without the arg info.

Reply:
Thank you, looks like an initialized local case to me indeed. One can reproduce it with something like:
```cs
int Write(bool a)
{
    Unsafe.SkipInit(out int b);
    if (a) b = 1;

    return b;
}
```
(A bit curious how Roslyn produces such locals for pattern matching. It can pessimize register allocation and range check elimination)